kernel/src/hil/display.rs

```diff
@@ -0,0 +1,334 @@
+    /// requesting it from the screen.
+    fn get_pixel_format(&self) -> ScreenPixelFormat;
+
+    /// Sets the video memory area.
+    ///
+    /// This function has to be called before the first call to the write function.
+    ///
+    /// Area's validity depends on the current pixel format:
+    /// - the first pixel of the first column must start on a byte boundary, and
+    /// - the last pixel of the last column must start on a byte boundary.
+    /// This is to reduce ambiguity regarding partial bytes.
+    ///
+    /// This will generate a `command_complete()` callback when finished.
+    ///
+    /// Return values:
+    /// - `Ok(())`: The write area is valid.
+    /// - `INVAL`: The parameters of the write area are not valid.
+    /// - `BUSY`: Unable to set the write area on the device.
+    fn set_write_area(&self, area: &Area) -> Result<(), ErrorCode>;
+
+    /// Sends a write command to write data in the selected video memory area.
+    /// When finished, the driver will call the `write_complete()` callback.
+    ///
+    /// Return values:
+    /// - `Ok(())`: Write is valid and will be sent to the screen.
+    /// - `RESERVE`: No write area reserved.
+    /// - `INVAL`: Write is invalid or length is wrong.
+    /// - `BUSY`: Another write is in progress.
+    fn write(&self, buffer: &'static mut [u8], len: usize) -> Result<(), ErrorCode>;
+
+    /// Applies all unapplied write commands.
+    ///
+    /// When finished, the driver shall call `ScreenClient::write_complete`.
+    /// Returns `BUSY` if another operation is in progress,
+    /// or `ENOSUPPORT` if the driver automatically applies all writes.
+    fn flush(&self) -> Result<(), ErrorCode> {
+        Err(ErrorCode::NOSUPPORT)
+    }
+
+    /// Set the object to receive the asynchronous command callbacks.
+    fn set_client(&self, client: Option<&'static dyn ScreenClient>);
+}
+
+/// Describes a display module with physical properties.
+///
+/// Must be implemented on top of a `FrameBuffer`.
+pub trait ScreenModule {
+    /// Sets the display brightness on a logarithmic scale
+    ///
+    /// Displays should implement this function for at least 0 and 1.
+    /// - 0 - completely no light emitted
+    /// - otherwise - on, set brightness to value
+    fn set_brightness(&self, brightness: usize) -> Result<(), ErrorCode>;
+
+    /// Controls the screen power supply.
+    ///
+    /// Use it to initialize the display device.
+    ///
+    /// Does not control backlight power (if applicable),
+    /// so call `set_brightness` to turn on/off the module completely.
+    ///
+    /// When finished, calls `ScreenClient::screen_is_ready`,
+    /// both when power was enabled and disabled.
+    fn set_power(&self, enabled: bool) -> Result<(), ErrorCode>;
+}
+
+/// A screen which has some extra processing capabilities.
+///
+/// Only available to drivers implementing FrameBuffer.
+pub trait FixedFunctionScreen {
```

Comment:
Why this name? What I tried to convey here is that this has fancy functions that most displays are not expected to have.

Reply:
`Display` - has only basic functions to actually display some content
`FrameBufferDisplay` - functions related to the framebuffer, like rotation, inversion. `FixedFunctionScreen`, at least for me, does not describe that. If you have any other names suggestions, I'm open to them.

Reply:
In https://github.com/tock/tock/pull/3047 I called it "FancyFunction" :) The problem with "FrameBuffer" is that Display is also only framebuffer functionality, because this HIL is only for framebuffer-based displays. Having the functionality of writing data to pixels presumes that. (As opposed to oscilloscope-like displays or plotters or printers.)

Reply:
You have a point, but we need a better name than `FixedFunction` or `Fancy`.